### ISSUE TITLE
Building from source: ffi-napi requires node >= 12.8

### ISSUE THREAD

**Issue by bernhardkaindl** [balena-io/balena-cli#2445]
On Ubuntu 20.04 an Node 10 or older errors like these will occur:
```
  CXX(target) Release/obj.target/ffi_bindings/src/ffi.o
In file included from /home/balena/balena-cli/node_modules/ffi-napi/node_modules/node-addon-api/napi.h:2725,
                 from ../src/ffi.h:2,
                 from ../src/ffi.cc:3:
/home/balena/balena-cli/node_modules/ffi-napi/node_modules/node-addon-api/napi-inl.h: In member function ‘T* Napi::Env::GetInstanceData()’:
/home/balena/balena-cli/node_modules/ffi-napi/node_modules/node-addon-api/napi-inl.h:471:24: error: there are no arguments to ‘napi_get_instance_data’ that depend on a template parameter, so a declaration of ‘napi_get_instance_data’ must be available [-fpermissive]
  471 |   napi_status status = napi_get_instance_data(_env, &data);
      |                        ^~~~~~~~~~~~~~~~~~~~~~
/home/balena/balena-cli/node_modules/ffi-napi/node_modules/node-addon-api/napi-inl.h:471:24: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
../src/ffi.cc: In static member function ‘static FFI::InstanceData* FFI::InstanceData::Get(Napi::Env)’:
../src/ffi.cc:11:24: error: ‘napi_get_instance_data’ was not declared in this scope; did you mean ‘napi_new_instance’?
   11 |   napi_status status = napi_get_instance_data(env, &d);
      |                        ^~~~~~~~~~~~~~~~~~~~~~
      |                        napi_new_instance
../src/ffi.cc: In function ‘Napi::Object BindingHook(Napi::Env, Napi::Object)’:
../src/ffi.cc:348:24: error: ‘napi_set_instance_data’ was not declared in this scope; did you mean ‘napi_new_instance’?
  348 |   napi_status status = napi_set_instance_data(
      |                        ^~~~~~~~~~~~~~~~~~~~~~
      |                        napi_new_instance
make: *** [ffi_bindings.target.mk:116: Release/obj.target/ffi_bindings/src/ffi.o] Error 1
make: Leaving directory '/home/balena/balena-cli/node_modules/ffi-napi/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/share/nodejs/node-gyp/lib/build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (events.js:198:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)
gyp ERR! System Linux 5.11.0-1019-oracle
gyp ERR! command "/usr/bin/node" "/usr/bin/node-gyp" "rebuild"
gyp ERR! cwd /home/balena/balena-cli/node_modules/ffi-napi
gyp ERR! node -v v10.19.0
gyp ERR! node-gyp -v v6.1.0
```
As mentioned in https://github.com/TooTallNate/NodObjC/pull/119 `api_napi_get_instance_data` requires node 12.8, in another bug report, it was said that ffi-napi 4.0.3 needs 12.18.3.

Please update the minimum version requirements here accordingly:
https://github.com/balena-io/balena-cli/blob/master/INSTALL-ADVANCED.md#npm-installation

Also prepend the instruction to run `npm install balena-cli --global --production --unsafe-perm` with `sudo` as the global installation to /usr requires root.

**Comment by pdcastro** [comment]
@bernhardkaindl, thank you for reporting this issue 👍  and sorry for the outdated details in the documentation. I have fixed it in PR #2446.

> in another bug report, it was said that ffi-napi 4.0.3 needs 12.18.3

I was able to build and run the CLI with Node.js v12.8.0, so I set the minimum requirement to v12.8.0. If you meant https://github.com/TooTallNate/NodObjC/pull/119#issuecomment-750797544, they mention an error in their Electron app while using Node.js v12.18.3, but they don't say that any error is resolved by using Node.js v12.18.3 or later.

> Also prepend the instruction to run `npm install balena-cli --global --production --unsafe-perm` with `sudo` as the global installation to `/usr` requires root.

But not when using `nvm`, right? The advanced installation instructions (npm installation) assume the use of `nvm`:

> https://github.com/balena-io/balena-cli/blob/master/INSTALL-ADVANCED.md#linux-or-wsl-windows-subsystem-for-linux
> The `curl` command line above uses [nvm](https://github.com/nvm-sh/nvm/blob/master/README.md#install--update-script) to install Node.js, instead of using `apt-get`. Installing Node.js through `apt-get` is a common source of problems from permission errors to conflict with other system packages, and therefore not recommended.

Also, 'sudo' does not even exist on Windows, so the instructions would have to add further system-specific exceptions and verbosity...

**Comment by bernhardkaindl** [comment]
@pdcastro Excellent! Many thanks for the fix and for verifying to version, and the additional explanation, now I got it.

Being a total Noob in the `nodejs` world, I didn't know a bit about `nvm`, and due to the similarity between `nvm` and `npm` and `npm install` and `nvm install 12` I was actually running `npm install 12` which of course didn't do what was expected, and ended up installing the nodejs-12 repo debian packages, with all the drawbacks which it mentions below.

Maybe it's good to expand the explanation a bit for people trying to build from source despite not knowing about npm and nvm?
This would also apply to users of ARM architectures, Apple M1 or ARM64 servers in the cloud.

Maybe like this, above the preformatted command block, to put there as a stronger recommendation: 

> **Recommendation**: It is strongly preferred do install nodejs 12 using the Node version manager [nvm](https://github.com/nvm-sh/nvm/blob/master/README.md#install--update-script) instead of `apt-get`, as this does not replace the nodejs packages on your system and does not require root privileges.
> Replacing the system-wide `nodejs`installation which consists of a huge amount of packages with `apt-get` is a common source of problems from permission errors to conflict with other system packages, and therefore not recommended.

Would you merge a PR for this, or would you like to do it?
Many thanks, Bernhard
